pkg/telemetry/events.go

```diff
@@ -0,0 +1,360 @@
+package telemetry
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/livekit/livekit-server/pkg/telemetry/prometheus"
+	"github.com/livekit/protocol/livekit"
+	"github.com/livekit/protocol/logger"
+	"github.com/livekit/protocol/utils"
+	"github.com/livekit/protocol/webhook"
+)
+
+func (t *telemetryService) NotifyEvent(ctx context.Context, event *livekit.WebhookEvent) {
+	if t.notifier == nil {
+		logger.Warnw("failed to notify webhook", errors.New("no notifier"), "event", event.Event)
+		return
+	}
+
+	event.CreatedAt = time.Now().Unix()
+	event.Id = utils.NewGuid("EV_")
+
+	t.webhookPool.Submit(func() {
+		if err := t.notifier.Notify(ctx, event); err != nil {
+			logger.Warnw("failed to notify webhook", err, "event", event.Event)
+		}
+	})
+}
+
+func (t *telemetryService) RoomStarted(ctx context.Context, room *livekit.Room) {
+	t.enqueue(func() {
+		t.NotifyEvent(ctx, &livekit.WebhookEvent{
+			Event: webhook.EventRoomStarted,
+			Room:  room,
+		})
+
+		t.SendEvent(ctx, &livekit.AnalyticsEvent{
+			Type:      livekit.AnalyticsEventType_ROOM_CREATED,
+			Timestamp: &timestamppb.Timestamp{Seconds: room.CreationTime},
+			Room:      room,
+		})
+	})
+}
+
+func (t *telemetryService) RoomEnded(ctx context.Context, room *livekit.Room) {
+	t.enqueue(func() {
+		t.NotifyEvent(ctx, &livekit.WebhookEvent{
+			Event: webhook.EventRoomFinished,
+			Room:  room,
+		})
+
+		t.SendEvent(ctx, &livekit.AnalyticsEvent{
+			Type:      livekit.AnalyticsEventType_ROOM_ENDED,
+			Timestamp: timestamppb.Now(),
+			RoomId:    room.Sid,
+			Room:      room,
+		})
+	})
+}
+
+func (t *telemetryService) ParticipantJoined(
+	ctx context.Context,
+	room *livekit.Room,
+	participant *livekit.ParticipantInfo,
+	clientInfo *livekit.ClientInfo,
+	clientMeta *livekit.AnalyticsClientMeta,
+) {
+	t.enqueue(func() {
+		prometheus.IncrementParticipantJoin(1)
+		prometheus.AddParticipant()
+
+		newWorker := newStatsWorker(
+			ctx,
+			t,
+			livekit.RoomID(room.Sid),
+			livekit.RoomName(room.Name),
+			livekit.ParticipantID(participant.Sid),
+			livekit.ParticipantIdentity(participant.Identity),
+		)
+		t.workersMu.Lock()
+		var free = false
+		for idx, worker := range t.workers {
```

Comment:
Not necessarily something to do now, but it would be nice to refactor the code to insert/delete/lookup entries in this workersIdx/workers datastructure pair in its own object.

Reply:
I'm not really sure why it isn't just a map

Reply:
I think I originally wrote it similar to https://github.com/livekit/livekit/blob/master/pkg/sfu/downtrackspreader.go#L93 for parallelization, but that got lost at some point. I'll change it to a map